### DISCUSSION TITLE
native css experiments, import assertions and/or ES modules

### DISCUSSION THREAD

**Discussion by rcherny** [webpack/webpack#18369] [Q&A]
I was toying with using the Native CSS support -- quite excited to leverage this (#14893). I have a combination of statements in the code previously supported by the `css-loader` and `mini-css-extract-plugin`.

Some are immediate loads when the parent JS module is loaded

```js
// parent js
import('./app/app1.js');
// app1.js
import './styles/file1.scss';
```
and some are dynamic imports (on demand):

```js
() => import('./styles/file2.scss';
```

Once I've enabled the `experiments.css` support — I'm realizing that this, combined with `output.chunkFormat: 'module';` and `experiments.outputModule: true` ... these files are no longer fetched!

I end up with errors like:

```shell
Uncaught (in promise) TypeError: e[n] is not a function
```

I'm realizing I may need to switch to import assertions (`with { type: "css" }`) to be in support of "web standards" for the modules.

I can do this natively, of course, but I need the final urls, chunks, and generated filenames for the CSS files to do this in production.

I see on the #14893 ticket that import assertion (Task B, probably with a change to `with`) is still unimplemented. The native support is great since with the `css-loader` and `mini-css` you end up with the extra JS files for CSS files that aren't actually necessary.

Does anyone know when that will land? Any alternatives to avoid generating a JS file for these "modules" that aren't actually JS modules at this point? Even if I had an import statement I could use in the JS file that would be updated and include the final URL, I could then `fetch()` the CSS file. I tried `new URL()` but it didn't work and -- yup, that task D is still on the not implemented list 😅 

Maybe I'm just too early? Maybe there's a feature (via insert?) that I'm missing in `mini-css-extract` that will allow me to parse these CSS links and get their URL? I don't even need to use the CSS members in the JS -- this is just on-demand loading of the CSS due to routing and file size limitations. Today it works fine but the extra requests is a bummer.

So:
1. either I need to just stick with the non-experiment version of css support with `css-loader` and `mini-css` (for now)
2. or is there a way to get this working with the `experiments.css` + `chunkFormat: module`? — incidentally, I've seen the `basic-esm-target-web` test case, but I haven't gotten that working for me, personally to date

I'm hoping someone in the community or even @alexander-akait may be able to point me in the right direction.

**Comment by rcherny** [comment]
Here's a (somewhat noisy) test repo with what i'm talking about. If you compare the main branch with the `webpack.exp-css-on-demand.config.js` with the branch `css-on-demand-esm` you can see the difference in behavior. 

https://github.com/rcherny/webpack-test-cases

**Reply by alexander-akait** [reply]
Sorry for delay

> Uncaught (in promise) TypeError: e[n] is not a function

Looks like you forgot to include CSS files in your HTML document

**Reply by alexander-akait** [reply]
We are working on `with` (will be in the next release) support instead assert, also in our plans:
- support `new URL("./style.css", import.meta.url)`, it will compile CSS and return URL to compiled file
- `with { type: "css" }` will work by spec

**Reply by alexander-akait** [reply]
And yes, CSS should work with module output, do you try the latest version, we have tests for this and I think I fixed it recently - https://github.com/webpack/webpack/commit/2779470f89da0df7ce839e53aa42a57a1e2a323a

**Reply by alexander-akait** [reply]
Oh, I see, we still not publish it, release will be soon, we want to finish couple PRs before release and everything will work
